Add a test case InstallPackageToWebSiteFromUI

v-luzh opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes:

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->
**Add a test case InstallPackageToWebSiteFromUI: install different package into a bulk of C# ASP Web Site projects.**
## PR Checklist

- [ ] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

v-luzh committed:
Add a test case InstallPackageToWebSiteFromUI: Install different package to a bulk of C# ASP Web Site Projects

ghost commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/signed)](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=4769) <br/>All CLA requirements met.

nkolev92 commented:
Hey @v-luzh 

Is the idea that these tests eventually make it into dev? I'm trying to understand whether we should review it with the standards in https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md

v-luzh commented:
Hi @nkolev92, we're trying to convert some manual testcases to automation. Currently, we cannot run apex testcases locally, so we need to submit PR and validate those testcases into privateDev. This PR is not ready for review since the testcase still have some issues we need to fix.

nkolev92 commented:
Thanks for the clarification. 

I'd recommend moving this to a draft then. 

![image](https://user-images.githubusercontent.com/2878341/187270314-f60d8a3f-bb33-4dd6-9856-c2c41a81f08d.png)

v-luzh commented:
Hi @nkolev92, it's OK for us to move it to draft. And could you please help to review the new test case: InstallPackageToWebSiteFromUI. We try to install different package with different version into Web Site template. Seems the target framework is not found in this PR.

nkolev92 commented:
are these the website projects with a framework or the ones that are just a project folder? Can you try not setting a framework at all?

I'm not very proficient in these tests either.

v-luzh commented:
Hi @nkolev92, but the corresponding test cases need to create the Web Site project with the target framework 4.8 or 4.5. And we can find the v48 in the "ProjectTargetFramework" as below.
![image](https://user-images.githubusercontent.com/20632625/187819326-a9554602-bad2-4043-97cc-7ca09986876d.png)

Any possible that all the Web Site projects workloads are not installed into VS before the build 6.4.0.296 running at all? The web site workloads are not in the default workloads when installing VS manually. We check the "Additional Project Templates" in the optional of "ASP.NET and web development" workload before every manual testing as below.
![image](https://user-images.githubusercontent.com/20632625/187821302-a5f903c9-28a5-4ba3-8718-8a8f17f0a6ec.png)

nkolev92 commented:
> Any possible that all the Web Site projects workloads are not installed into VS before the build 6.4.0.296 running at all? 

Could be. 

cc @NuGet/nuget-client Anyone can help answer Jamie's question?

zivkan commented:
This is the vsconfig that is installed for our E2E and Apex tests: https://github.com/NuGet/NuGet.Client/blob/dev/build/tests.dev17.vsconfig

I don't know what the "additional project templates" component id is, so I'm not sure if it's already selected.  The file is intentionally minimal, because the more components that are installed, the slower it is to install and therefore the slower the pipeline is to complete. Looking at this PR's last build, the Test Machine Deployment jobs already took 41-43 minutes. Anyway, feel free to add the additional templates to the vsconfig file (you can use the VS Installer's Export functionality), but please only add that, and don't tell it to install all components and workloads.  We're also working on implementing "build to build upgrades" which should significantly reduce the test machine deployment time, at which time maybe we can install all workloads and components, but for now we're trying to be minimal.

v-luzh commented:
Hi @zivkan, thanks for your update. After investigating, the C# Web Site workload was not installed into the VS for this run. And I have gotten the component name we should add into the vsconfig file. Shall we update the file (https://github.com/NuGet/NuGet.Client/blob/dev/build/tests.dev17.vsconfig) directly and re-run this case again?

zivkan commented:
Yes, you can update the file in the branch for this PR, then when you run CI on the latest commit, it will use the changed vsconfig file.

v-luzh commented:
@zivkan, when we trying to update the vsconfig file in the branch for this PR, a [minimal vsconfig file](https://github.com/NuGet/NuGet.Client/blob/dev-TestTeam-Automation/.vsconfig) shows instead of the one you mentioned above. Compare the vsconfig file I exported from the VSInstaller with the minimal vsconfig file we used in the dev-TestTeam-Automation branch, there are a lot of gaps. If we want to add more test cases moving forward, we should add a lot components into the minimal vsconfig file, because our cases are all End-To-End Test not Unit Test. Does that make sense? Or should we update the vsconfig file you mentioned (only need to add 3 lines) and use it in dev-TestTeam-Automation branch?

zivkan commented:
The `.vsconfig` at the repository root is the minimal to build NuGet, not to run tests (think of people who do not work for Microsoft, and are trying to contribute a small bug fix). It's also not used by CI tests. The `build/tests.dev17.vsconfig` that I mentioned earlier is the only one used by CI. This means that people who are writing/debugging tests and want to be able to run those tests locally will need to manually import the `build/tests.dev17.vsconfig` file in the VS Installer, rather than relying on Visual Studio's automatic detection of the `.vsconfig` file in the same directory as the `NuGet.sln` file.

v-luzh commented:
Thanks, @zivkan, got it. So we will update this vsconfig file: https://github.com/NuGet/NuGet.Client/blob/dev-TestTeam-Automation/build/tests.dev17.vsconfig, and then re-run the PR for the new test case.

nkolev92 commented:
Moving this to draft so that it doesn't appear in our to review queries. 

The CI can still be triggered as normal.

v-luzh committed:
Update tests.dev17.vsconfig

Add SDK 4.8.1 and AdditionalWebProjectTemplates to involve the C# Web Sites Application (.NET Framework 4.8.1).

v-luzh commented:
Hi @nkolev92, we updated the https://github.com/NuGet/NuGet.Client/blob/dev-TestTeam-Automation/build/tests.dev17.vsconfig, so the PR is ready for review now. Could you please help to review the case: InstallPackageToWebSiteFromUI? The scenario is from Daily Manual Test cases-[Web Sites](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc={8a3c6685-1661-47f4-8807-b765ca62aea1}&action=edit&wd=target%28Manual%20Walk%20Through.one%7C12a48230-138f-499b-8b9f-ea2a2ca54496%2FWeb%20Site%20%28Legacy%5C%29%7C91c7d59d-8afa-4697-9ed0-92d8aacbbbcb%2F%29&wdorigin=703).  We want to install different package into different Web Site project in PM UI through running this test case.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/ProjectDefination.cs:
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
FIlename has a typo, `ProjectDefinition`.

v-luzh replied:
Thanks for your review. Sorry, we will fix all the issues together later.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/ProjectDefination.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace NuGet.Tests.Apex.NuGetEndToEndTests
+{
+    internal class WebSiteProjectItems
```
Filename should match the class name.

v-luzh replied:
Thanks for your review. Sorry, we will fix all the issues together later.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/ProjectDefination.cs:
```diff
@@ -0,0 +1,41 @@
+    internal class WebSiteProjectItem
+    {
+        public ProjectTemplate Template { get; set; }
+        public string ProjectName { get; set; }
```
Name doesn't seem to be used.

v-luzh replied:
Thanks for your review. Sorry, we will fix all the issues together later. We want to save the solution as the ProjectName and reopen it at the beginning of next new automation case: UpdatePackageInExistedWebSiteProjectFromUI.

nkolev92 replied:
No worries. Take your time and let us know when it's ready.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs:
```diff
@@ -70,6 +71,35 @@ public void InstallPackageFromUI()
+
+            WebSiteProjectItems websiteProject = new WebSiteProjectItems();
+
+            foreach (WebSiteProjectItem item in websiteProject.InitProjectItems())
```
Given that we're testing multiple projects, we should create a `Theory`, and provide the data instead of looping over all the templates.

Here's an example: https://github.com/NuGet/NuGet.Client/blob/d22c6743bf6237c39b98d5d680e06e6c33e97f1c/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NetCoreProjectTestCase.cs#L195-L197

v-luzh replied:
Thanks for your suggestion. We are investigating how to fix this problem with `Theory` and fix all the issues together later.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/ProjectDefination.cs:
```diff
@@ -0,0 +1,41 @@
+            _websiteProjectItems.Add(new WebSiteProjectItem() { Template=ProjectTemplate.WebSite, ProjectName="WebSite", TargetFramework=ProjectTargetFramework.V48, PackageName="Log4Net.Async", PackageVersion="2.0.3" });
+            _websiteProjectItems.Add(new WebSiteProjectItem() { Template=ProjectTemplate.WebSiteWCF, ProjectName="WebSiteWCF", TargetFramework=ProjectTargetFramework.V48, PackageName="log4net.Ext.Json", PackageVersion="2.0.9.1" });
+            _websiteProjectItems.Add(new WebSiteProjectItem() { Template=ProjectTemplate.WebSiteDynamicDataEntityFramework, ProjectName="WebSiteEntities", TargetFramework=ProjectTargetFramework.V48, PackageName="log4net", PackageVersion="2.0.13" });
+            _websiteProjectItems.Add(new WebSiteProjectItem() { Template=ProjectTemplate.WebSiteDynamicDataLinqToSql, ProjectName="WebSiteLinq", TargetFramework=ProjectTargetFramework.V45, PackageName="Log4Net.Async", PackageVersion="2.0.3" });
```
Any particular reason why we're testing net45? I see the rest of the templates are net48, so I was wondering if there's any particular reason for that.

v-luzh replied:
Hi @nkolev92, for the last project-**ASP.NET Dynamic Data Linq to SQL Web Site**, it only support .NET Framework 4.5.2 and the framework cannot be changed when creating the project manually as below.
![image](https://user-images.githubusercontent.com/20632625/190295930-2acaab55-009c-407b-9744-6c68ffc8c67f.png)

nkolev92 replied:
Thanks for the explanation. Makes sense.  Might be a good idea to add a small comment in the test case declaration.

v-luzh committed:
Update tests.dev17.vsconfig

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.
